rspace/src/main/scala/coop/rchain/rspace/RSpace.scala

```diff
@@ -153,121 +189,179 @@ class RSpace[F[_], C, P, E, A, R, K] private[rspace] (
-          @tailrec
-          def extractProduceCandidate(
-              groupedChannels: Seq[Seq[C]],
-              batChannel: C,
-              data: Datum[A]
-          ): Either[E, Option[ProduceCandidate[C, P, R, K]]] =
-            groupedChannels match {
-              case Nil => Right(None)
-              case channels :: remaining =>
-                val matchCandidates: Seq[(WaitingContinuation[P, K], Int)] =
-                  store.withTxn(store.createTxnRead()) { txn =>
-                    Random.shuffle(store.getWaitingContinuation(txn, channels).zipWithIndex)
-                  }
-                /*
-                 * Here, we create a cache of the data at each channel as `channelToIndexedData`
-                 * which is used for finding matches.  When a speculative match is found, we can
-                 * remove the matching datum from the remaining data candidates in the cache.
-                 *
-                 * Put another way, this allows us to speculatively remove matching data without
-                 * affecting the actual store contents.
-                 *
-                 * In this version, we also add the produced data directly to this cache.
-                 */
-                val channelToIndexedData: Map[C, Seq[(Datum[A], Int)]] = channels.map { (c: C) =>
-                  val as = store.withTxn(store.createTxnRead()) { txn =>
-                    Random.shuffle(store.getData(txn, Seq(c)).zipWithIndex)
-                  }
-                  c -> {
-                    if (c == batChannel) (data, -1) +: as else as
-                  }
-                }.toMap
-                extractFirstMatch(channels, matchCandidates, channelToIndexedData) match {
-                  case Left(e)                 => Left(e)
-                  case Right(None)             => extractProduceCandidate(remaining, batChannel, data)
-                  case Right(produceCandidate) => Right(produceCandidate)
-                }
-            }
-
-          extractProduceCandidate(groupedChannels, channel, Datum(data, persist, produceRef)) match {
-            case Left(e) => Left(e)
-            case Right(
-                Some(
-                  ProduceCandidate(
-                    channels,
-                    WaitingContinuation(patterns, continuation, persistK, consumeRef),
-                    continuationIndex,
-                    dataCandidates
-                  )
-                )
-                ) =>
-              produceCommCounter.increment()
+      for {
+        produceRef <- Produce.create(channel, data, persist, sequenceNumber).pure[F]
+        result <- produceLockF(channel) {
+                   /*
+                    * Find produce candidate
+                    *
+                    * Could also be implemented with a lazy `foldRight`.
+                    */
+                   type MaybeCandidate = Option[ProduceCandidate[C, P, R, K]]
+                   //@tailrec <- fix tailrec
```

Comment:
Does it matter that this function is tail recursive, since flatMap is trampolined?

Reply:
I believe after the refactor to tailrecM / iterateWhileM this is a stray comment - and no matter what monad we use (as long as it has a lawful tailrecM)